skglm/penalties/block_separable.py

```diff
@@ -153,3 +153,91 @@ def subdiff_distance(self, W, grad, ws):
+]
+
+
+@jitclass(spec_BlockSCAD)
+class BlockSCAD(BasePenalty):
+    """Block Smoothly Clipped Absolute Deviation.
+
+    Notes
+    -----
+    With W_j the j-th row of W, the penalty is:
+        pen(||W_j||) = alpha * ||W_j||               if ||W_j|| =< alpha
+                       (2 * gamma * alpha * ||W_j|| - ||W_j|| ** 2 - alpha ** 2) \
+                           / (2 * (gamma - 1))       if alpha < ||W_j|| < alpha * gamma
+                       (alpha **2 * (gamma + 1)) / 2 if ||W_j|| > gamma * alpha
+        value = sum_{j=1}^{n_features} pen(||W_j||)
+    """
+
+    def __init__(self, alpha, gamma):
+        self.alpha = alpha
+        self.gamma = gamma
+
+    def value(self, W):
+        """Compute the value of the SCAD penalty at W."""
+        n_features = W.shape[0]
+        norm_rows = np.sqrt(np.sum(W ** 2, axis=1))
+        value = np.full_like(norm_rows, (self.gamma + 1) * self.alpha ** 2 / 2.)
+        for j in range(n_features):
+            if norm_rows[j] <= self.alpha:
+                value[j] = self.alpha * norm_rows[j]
+            elif norm_rows[j] <= self.alpha * self.gamma:
+                value[j] = (
+                    2 * self.gamma * self.alpha * norm_rows[j] - norm_rows[j] ** 2
+                    - self.alpha ** 2) / (2 * (self.gamma - 1))
+        return np.sum(value)
+
+    def prox_1feat(self, value, stepsize, j):
+        """Compute the proximal operator of BlockSCAD."""
+        tau = self.gamma * self.alpha
+        x_1 = max(0, norm(value) - self.alpha * stepsize)
+        x_2 = ((self.gamma - 1) * norm(value) - stepsize * tau) / (
+            self.gamma - 1 - stepsize)
+        x_2 = np.abs(x_2)
+        x_3 = norm(value)
+
+        objs = np.zeros(3)
+        # TODO
+        objs[0] = 0.
+        objs[1] = 0.
+        objs[2] = 0.
+
+        idx_min = np.argmin(objs)
+        if idx_min == 0:
+            return norm(value) * x_1
+        elif idx_min == 1:
+            return norm(value) * x_2
+        else:
+            return norm(value) * x_3
+
+    def subdiff_distance(self, W, grad, ws):
+        """Compute distance of negative gradient to the subdifferential at W."""
+        subdiff_dist = np.zeros_like(ws, dtype=grad.dtype)
+        for idx, j in enumerate(ws):
+            norm_Wj = norm(W[j])
+            if not np.any(W[j]):
+                # distance of -grad_j to alpha * unit_ball
+                subdiff_dist[idx] = max(0, norm(grad[idx]) - self.alpha)
+            elif norm_Wj <= self.alpha:
+                # distance of -grad_j to alpha * W[j] / ||W[j]||
+                subdiff_dist[idx] = norm(grad[idx] + self.alpha * W[j] / norm_Wj)
+            elif norm_Wj <= self.gamma * self.alpha:
+                # distance of -grad_j to (alpha * gamma - ||W[j]||)
+                # / ((gamma - 1) * ||W[j]||) * W[j]
+                subdiff_dist[idx] = norm(grad[idx] + (
+                    (self.alpha * self.gamma - norm_Wj) / (norm_Wj * (self.gamma - 1))
+                ) * W[j])
```

Comment:
@mathurinm @Klopfe could you double check this case?

Reply:
Seems good to me!